ynr/apps/people/models.py

```diff
@@ -57,6 +64,92 @@ class PersonImage(models.Model):
     objects = PersonImageManager()
 
 
+class PersonIdentifier(TimeStampedModel):
+    """
+    This is a model for storing "identifiers" for a person.
+
+    The most simple case is a URL to another website.
+
+    In this case the "value" is "@democlub" (or it could be
+    "https"//twitter.com/democlub").
+
+    Twitter uses "vanity URLs" as well as internal identifiers so we can
+    store the internal ID that looks like "362837635" on the internal_identifier
+    field.
+
+    An identifier value or internal identifier doesn't have to be resolvable
+    over HTTP. For example a phone number or snapchat handle are both valid
+    values.
+
+    """
+
+    person = models.ForeignKey(
+        "people.Person", related_name="tmp_person_identifiers"
+    )
+    value = models.CharField(
+        max_length=800,
+        help_text="An identifier e.g a URL or username provided by a 3rd party",
+    )
+    internal_identifier = models.CharField(
+        max_length=800,
+        help_text="An optional internal identifier from the 3rd party",
+        null=True,
+    )
+    value_type = models.CharField(
+        max_length=100,
+        help_text="A label for the type of value e.g. 'Twitter', 'Person blog'",
+    )
+    extra_data = JSONField(
+        help_text="""For storing any additional data against this field. 
+                     Used by bots, not humans.""",
+        null=True,
+    )
+
+    objects = PersonIdentifierQuerySet.as_manager()
+
+    class Meta:
+        unique_together = (
+            ("person", "value"),
+            ("person", "internal_identifier", "value_type"),
+            # TODO: Remove this.
+            # At the moemnt the version history can't deal with more than one
+            # value per value_type. This prevents creating duplicates, and
+            # therefore means we'll be able to merge and revert people.
+            # This constraint should be removed when the version history can
+            # support the data model we want
+            ("person", "value_type"),
+        )
+        ordering = ("value_type", "-modified")
+
+    def __str__(self):
+        return "{}: {} ({})".format(self.person_id, self.value_type, self.value)
+
+    @property
+    def get_value_type_html(self):
+        return self.value_type.title()
```

Comment:
It seems like what we're trying to do here is present a nice human-readable title, but we're showing the user titles with underscores in them e.g:

![screenshot at 2018-11-05 13-02-10](https://user-images.githubusercontent.com/6025893/48000860-bd79ef00-e0fe-11e8-8c27-48bb2238e983.png)

Reply:
I've sort of hacked a fix for this with some calls to `replace` – Reason I've not done any more on this is that we'll need to think about how to display value type labels vs value type keys later on – I don't want to get too in to doing that just now only to undo something later.

It's possible we end up with something a bit like `ComplexPopoloFields` where the user facing label is defined in the database. This would allow for nicer management by power users.

This would still be an improvement on the current system because it wouldn't involve GFKs or splitting value types across models.